src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticService.cs

```diff
@@ -66,101 +65,111 @@ internal partial class DiagnosticService : IDiagnosticService
             }
         }
 
-        private void RaiseDiagnosticsUpdated(IDiagnosticUpdateSource source, DiagnosticsUpdatedArgs args)
+        private void RaiseDiagnosticsUpdated(IDiagnosticUpdateSource source, ImmutableArray<DiagnosticsUpdatedArgs> argsCollection)
         {
-            _eventListenerTracker.EnsureEventListener(args.Workspace, this);
+            Workspace? previousWorkspace = null;
+            foreach (var args in argsCollection)
+            {
+                if (args.Workspace != previousWorkspace)
```

Comment:
Here you explicitly handle possibility of args coming from different workspaces, while you assumed they were coming from the same workspace in a prior file. We want to be consistent and either assert at all places that args are from the same workspace OR allow and handle args coming from different workspaces.